### PR TITLE
chore(#9231): filter aggregate targets by period

### PR DESCRIPTION
**Pull request by Benmuiruri**
# Description

After the completion of https://github.com/medic/cht-core/pull/9267 which implements the actual filtering of the Aggregate Targets through the selection of the facilities in the radio buttons. 

This PR adds the feature of filtering by reporting period. This filtering will be available for all supervisors (With one facility or with multi-facility). 

<details>
<summary>Video demo and details of expected content on filtering</summary>

### The first half of the video shows user with multi-facilities

- Toggling the Health Facility on the sidebar changes the label of the LHS list from `Ben's Health Facility` to `Second Facility`
Screenshot 2024-08-06 at 12 04 39
- Toggling the Reporting Period on the sidebar changes the label of the LHS list from `Ben's Health Facility` to `Ben's Health Facility - {{ Previous month }}` when you select `Last month`
Screenshot 2024-08-06 at 12 06 41

### The second hald of the video shows user with one facility 

- When user has one facility, it does not show the facility name on the LHS and we don't show the Health Facility filter
- When user has one facility, it shows the  `{{ Previous month }}` on the LHS when you select `Last month`

https://github.com/user-attachments/assets/2616ddc6-3870-4f91-b566-b02b41120b2a

Mobile View 

https://github.com/user-attachments/assets/547c170b-6d72-4398-9dee-ad0a6c433127

</details>

# Code review checklist

- [ ] Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- [ ] Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- [ ] Tested: Unit and/or e2e where appropriate
- [ ] Internationalised: All user facing text
- [ ] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.

# Compose URLs
If Build CI hasn't passed, these may 404:

* [Core](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:filter-aggregate-period/docker-compose/cht-core.yml)
* [CouchDB Single](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:filter-aggregate-period/docker-compose/cht-couchdb.yml)
* [CouchDB Cluster](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:filter-aggregate-period/docker-compose/cht-couchdb-clustered.yml)

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

**Comment by Benmuiruri** [comment]
Hi @latin-panda here is an summary of what I did and what's left

**What I have done**

- Display the filter button to all supervisors 
- Show facility filter options in the sidebar only to multi-facility users 
- Show reporting period filter options in the sidebar only to all users  
- Switching reporting period fetches the aggregate target by reporting period. I added this [log](https://github.com/medic/cht-core/pull/9283/files#:~:text=console.log(%27tag%27%2C%20tag)%3B) as well to check the tag 

**What is left to do**

- [ ] Add translations for reporting [period text](https://github.com/medic/cht-core/pull/9283/files#diff-a9b0c838bb2a123ccb22e7b4838124c757bdbed6535d76c1a0fb9a898d008fba:~:text=%27analytics.target.aggregates.reporting_period%27) 
- [ ] Add translation for [reporting radio ](https://github.com/medic/cht-core/pull/9283/files#diff-a9b0c838bb2a123ccb22e7b4838124c757bdbed6535d76c1a0fb9a898d008fba:~:text=%27analytics.target.aggregates.this_month%27)
- [ ] change name of `filter-option-place` [class](https://github.com/medic/cht-core/blob/98c079e29126684680cfffd2782e8010d674211e/webapp/src/ts/modules/analytics/analytics-target-aggregates-sidebar-filter.component.html#L39) to one more generic like  `filter-radio-option` 
- [ ] Change the class name in `sidebar-filter.less` also
- [ ] Consideration for how `Facility Label breadcrumb` and `Reporting period period breadcrumb` will display
- [ ] Update [getCurrentTargetDoc](https://github.com/medic/cht-core/blob/98c079e29126684680cfffd2782e8010d674211e/webapp/src/ts/services/target-aggregates.service.ts#L337) maybe like I did it for the [Docs](https://github.com/medic/cht-core/pull/9283/files#:~:text=private-,fetchLatestTargetDocs,-(settings%2C) to update respond when reporting period is toggled. 
- [ ] Submit reports in previous months and toggle reporting period to validate correctness of data in aggregate list
- [ ] Unit tests and possible e2e test

File: webapp/src/ts/modules/analytics/analytics-target-aggregates-sidebar-filter.component.html [hunk 1]
```diff
@@ -31,6 +32,35 @@
               </div>
             </ng-container>
         </div>
+
+        <!-- 2. Reporting Period -->
+        <div class="filter-options-container">
```
Comment (latin-panda) [inline webapp/src/ts/modules/analytics/analytics-target-aggregates-sidebar-filter.component.html:37]: I think this should be in a separate accordion

**Comment by latin-panda** [comment]
@Benmuiruri I didn't have time to do much here. This is what I did: 
1. I fixed conflicts, and some things changed to reflect the default values in the indicator text. 
2. Added enums for the `current` and `previous` to not hardcode 
3. Added the selected reporting period to the indicator text - but still needs translations

Screenshot 2024-07-30 at 12 53 26 PM

File: webapp/src/ts/services/target-aggregates.service.ts [hunk 2]
```diff
@@ -49,12 +50,21 @@ export class TargetAggregatesService {
...
+  }
+
   /**
    * Every target doc follows the _id scheme `target~<interval_tag>~<contact_uuid>~<user_id>`
    * In order to retrieve the latest target document(s), we compute the current interval <interval_tag>
    */
-  private fetchLatestTargetDocs(settings) {
-    const tag = this.getCurrentIntervalTag(settings);
+  private fetchLatestTargetDocs(settings, reportingPeriod?) {
+    const tag = reportingPeriod === 'previous'
```
Comment (latin-panda) [inline webapp/src/ts/services/target-aggregates.service.ts:64]: Use enum here

**Comment by Benmuiruri** [comment]
Thanks for rebasing and the progress you made, I will continue with it.

File: webapp/src/ts/services/target-aggregates.service.ts [hunk 3]
```diff
@@ -49,12 +53,21 @@ export class TargetAggregatesService {
...
   /**
    * Every target doc follows the _id scheme `target~<interval_tag>~<contact_uuid>~<user_id>`
    * In order to retrieve the latest target document(s), we compute the current interval <interval_tag>
    */
-  private fetchLatestTargetDocs(settings) {
-    const tag = this.getCurrentIntervalTag(settings);
+  private fetchLatestTargetDocs(settings, reportingPeriod?: ReportingPeriod) {
+    const tag = reportingPeriod === ReportingPeriod.PREVIOUS
+      ? this.getPreviousIntervalTag(settings)
+      : this.getCurrentIntervalTag(settings);
```
Comment (latin-panda) [inline webapp/src/ts/services/target-aggregates.service.ts:69]:
```suggestion
    const tag = this.isPreviousPeriod(reportingPeriod)
      ? this.getPreviousIntervalTag(settings)
      : this.getCurrentIntervalTag(settings);
```

File: webapp/src/ts/services/target-aggregates.service.ts [hunk 4]
```diff
@@ -295,11 +308,23 @@ export class TargetAggregatesService {
     return !facilityIds || facilityIds.length > 0;
   }
 
-  getAggregates(facilityId?) {
-    return this.ngZone.runOutsideAngular(() => this._getAggregates(facilityId));
+  getReportingMonth(reportingPeriod) {
+    return this.settingsService
+      .get()
+      .then(settings => {
+        const tag = reportingPeriod === ReportingPeriod.PREVIOUS
```
Comment (latin-panda) [inline webapp/src/ts/services/target-aggregates.service.ts:315]: This code is repeated, so better make a private `getTag` function.

Please add unit tests for the new functions

File: webapp/src/ts/services/target-aggregates.service.ts [hunk 5]
```diff
@@ -358,6 +383,9 @@ export interface AggregateTarget extends Target {
   isPercent: boolean;
   progressBar: boolean;
   facility?: string;
+  reportingPeriod?: ReportingPeriod;
+  reportingMonth?: string;
+  lineage?: (string | ReportingPeriod)[];
```
Comment (latin-panda) [inline webapp/src/ts/services/target-aggregates.service.ts:388]:
```suggestion
  filtersToDisplay?: (string | ReportingPeriod)[];
```

**Review by latin-panda — CHANGES_REQUESTED** [review]
Exciting to see this feature almost ready for release!! 

I've added some comments below

File: shared-libs/calendar-interval/test/index.js [hunk 6]
```diff
@@ -252,6 +252,35 @@ describe('CalendarInterval', () => {
     });
   });
 
+  describe('getPrevious', () => {
```
[the remark this answers is not in the document]
Comment (Benmuiruri) [reply shared-libs/calendar-interval/test/index.js:255]: I dont understand this one `when reference date is sent (past date, and future date)` please clarify.

[the remark this answers is not in the document]
Comment (latin-panda) [reply shared-libs/calendar-interval/test/index.js:255]: The second parameter of this function is reference date, which defaults to "now" (`moment()`). What is the expected behavior when passing to that parameter a date in the past or the future?

File: webapp/src/ts/modules/analytics/analytics-target-aggregates-detail.component.ts [hunk 7]
```diff
@@ -110,4 +115,14 @@ export class AnalyticsTargetAggregatesDetailComponent implements OnInit, OnDestr
...
   }
+
+  getReportingPeriodText(): string {
+    const aggregateSubtitle = this.translateService.instant(this.selected.subtitle_translation_key);
+
+    if (this.selected.reportingPeriod === ReportingPeriod.CURRENT) {
+      return aggregateSubtitle;
+    }
+
+    return `${aggregateSubtitle} (${this.selected.reportingMonth})`;
```
[the remark this answers is not in the document]
Comment (Benmuiruri) [reply webapp/src/ts/modules/analytics/analytics-target-aggregates-detail.component.ts:126]: Excellent question. I had not considered that. 

How about having a default text. I'm thinking updating `getTargetAggregates` in  `webapp/src/ts/modules/analytics/analytics-target-aggregates.component.ts` to this. 

```
const reportingMonth = await this.targetAggregatesService
        .getReportingMonth(reportingPeriod) || this.translateService.instant('admin.pagination.prev');
```

when `reportingMonth` fails we would display `This Month (Previous)` what do you think of the approach?

[the remark this answers is not in the document]
Comment (latin-panda) [reply webapp/src/ts/modules/analytics/analytics-target-aggregates-detail.component.ts:126]: en: `Up to last month` 
es: `Hasta el mes pasado`
id: `Sampai bulan lalu`

File: webapp/src/ts/modules/analytics/analytics-target-aggregates-sidebar-filter.component.html [hunk 8]
```diff
@@ -1,41 +1,59 @@
...
+      </p>
       <a class="sidebar-close fa fa-times" (click)="toggleSidebarFilter()"></a>
     </div>
 
     <div class="sidebar-body">
       <mat-accordion multi>
+        <!-- Facility -->
+        <div class="filter-options-container" *ngIf="userFacilities.length > 1">
+          <span class="filter-options-title"> {{ facilityFilterLabel | translate
+            }} </span>
```
Comment (latin-panda) [inline webapp/src/ts/modules/analytics/analytics-target-aggregates-sidebar-filter.component.html:18]:
```suggestion
          <span class="filter-options-title"> 
            {{ facilityFilterLabel | translate }} 
          </span>
```

File: webapp/src/ts/modules/analytics/analytics-target-aggregates-sidebar-filter.component.html [hunk 9]
```diff
@@ -1,41 +1,59 @@
...
+          <ng-container *ngFor="let facility of userFacilities">
+            <div class="filter-radio-option">
+              <label class="filter-option-wrapper">
+                <input type="radio" name="facility" [value]="facility"
+                  class="filter-option-radio"
+                  [checked]="facility._id === selectedFacility._id"
+                  (change)="fetchAggregateTargetsByFacility(facility)">
```
Comment (latin-panda) [inline webapp/src/ts/modules/analytics/analytics-target-aggregates-sidebar-filter.component.html:25]:
```suggestion
                <input 
                  type="radio" 
                  name="facility" 
                  [value]="facility"
                  class="filter-option-radio"
                  [checked]="facility._id === selectedFacility._id"
                  (change)="fetchAggregateTargetsByFacility(facility)">
```

File: webapp/src/ts/modules/analytics/analytics-target-aggregates-sidebar-filter.component.html [hunk 10]
```diff
@@ -1,41 +1,59 @@
...
-                    [value]="facility"
-                    class="filter-option-radio"
-                    [checked]="facility._id === selectedFacility._id"
-                    (click)="fetchAggregateTargets(facility)">
-                  <span class="filter-option-label">{{ facility.name }}</span>
-                </label>
-              </div>
-            </ng-container>
+          <span class="filter-options-title">{{
+            'analytics.target.aggregates.reporting_period' | translate }}</span>
```
Comment (latin-panda) [inline webapp/src/ts/modules/analytics/analytics-target-aggregates-sidebar-filter.component.html:37]:
```suggestion
          <span class="filter-options-title">{{ 'analytics.target.aggregates.reporting_period' | translate }}</span>
```

File: webapp/src/ts/modules/analytics/analytics-target-aggregates-sidebar-filter.component.html [hunk 11]
```diff
@@ -1,41 +1,59 @@
...
+          <div class="filter-radio-option"
+            *ngFor="let period of reportingPeriods">
```
Comment (latin-panda) [inline webapp/src/ts/modules/analytics/analytics-target-aggregates-sidebar-filter.component.html:39]:
```suggestion
          <div class="filter-radio-option" *ngFor="let period of reportingPeriods">
```

File: webapp/src/ts/modules/analytics/analytics-target-aggregates-sidebar-filter.component.html [hunk 12]
```diff
@@ -1,41 +1,59 @@
...
+            <label class="filter-option-wrapper">
+              <input type="radio" name="reporting-period"
+                class="filter-option-radio" [value]="period.value"
+                [(ngModel)]="selectedReportingPeriod"
+                (change)="fetchAggregateTargetsByReportingPeriod()">
```
Comment (latin-panda) [inline webapp/src/ts/modules/analytics/analytics-target-aggregates-sidebar-filter.component.html:44]:
```suggestion
              <input 
                type="radio" 
                name="reporting-period"
                class="filter-option-radio" [value]="period.value"
                [(ngModel)]="selectedReportingPeriod"
                (change)="fetchAggregateTargetsByReportingPeriod()">
```

File: webapp/src/ts/modules/analytics/analytics-target-aggregates-sidebar-filter.component.html [hunk 13]
```diff
@@ -1,41 +1,59 @@
...
+              <span class="filter-option-label">{{ period.label | translate
+                }}</span>
```
Comment (latin-panda) [inline webapp/src/ts/modules/analytics/analytics-target-aggregates-sidebar-filter.component.html:46]:
```suggestion
              <span class="filter-option-label">
                {{ period.label | translate }}
              </span>
```

File: webapp/src/ts/modules/analytics/analytics-target-aggregates-sidebar-filter.component.html [hunk 14]
```diff
@@ -1,41 +1,59 @@
...
+          </div>
         </div>
       </mat-accordion>
     </div>
 
     <div class="sidebar-footer">
-      <button type="button" class="btn btn-primary" (click)="toggleSidebarFilter()">{{ 'reports.sidebar.filter.submit' | translate }}</button>
+      <button type="button" class="btn btn-primary"
+        (click)="toggleSidebarFilter()">{{ 'reports.sidebar.filter.submit' |
+        translate }}</button>
```
Comment (latin-panda) [inline webapp/src/ts/modules/analytics/analytics-target-aggregates-sidebar-filter.component.html:56]:
```suggestion
      <button type="button" class="btn btn-primary" (click)="toggleSidebarFilter()">
        {{ 'reports.sidebar.filter.submit' | translate }}
      </button>
```

File: webapp/src/ts/modules/analytics/analytics-target-aggregates-sidebar-filter.component.ts [hunk 15]
```diff
@@ -74,8 +85,17 @@ export class AnalyticsTargetAggregatesSidebarFilterComponent implements OnInit,
...
   }
+
+  fetchAggregateTargetsByReportingPeriod() {
+    this.reportingPeriodSelectionChanged.emit(this.selectedReportingPeriod);
+  }
+}
+
+export enum ReportingPeriod {
+  CURRENT = 'targets.this_month.subtitle',
+  PREVIOUS = 'targets.last_month.subtitle'
```
Comment (latin-panda) [inline webapp/src/ts/modules/analytics/analytics-target-aggregates-sidebar-filter.component.ts:100]: Usually, enums use simple words that don't change much and are easy to read. But labels (translation keys) don't seem a good fit for this, feels unexpected for enums. 

Is there a reason? Can we keep simple `current` and `previous` values? That would give `readonly reportingPeriods` (L20) more reasons to exist.

File: webapp/src/ts/modules/analytics/analytics-target-aggregates.component.html
Comment (latin-panda) [inline webapp/src/ts/modules/analytics/analytics-target-aggregates.component.html:1]: Can you please double-check this file? There are some breaks in weird places.

File: api/resources/translations/messages-es.properties [hunk 16]
```diff
@@ -638,7 +639,7 @@ email.invalid = Dirección de correo electrónico no válida.
 empty = El mensaje esta en blanco, por favor reenvielo. Si continua teniendo problemas, informe a su supervisor.
 enketo.constraint.invalid = Valor no permitido
 enketo.constraint.required = Este campo es obligatorio
-enketo.drawwidget.annotation = anotaci�n
+enketo.drawwidget.annotation = anotación
```
Comment (latin-panda) [inline api/resources/translations/messages-es.properties:642]: Ben, were you able to find a solution for these?

**Review by latin-panda — CHANGES_REQUESTED** [review]
@Benmuiruri almost ready! I left some questions below and comments
I still need to do another round tomorrow, so I might send more feedback

File: api/resources/translations/messages-es.properties (hunk 16 above)
Comment (Benmuiruri) [reply api/resources/translations/messages-es.properties:642]: Yes before the rebase I had fixed it. 

Here it is now as you see it. 

Screenshot 2024-08-06 at 18 24 18

I've pushed and now I see it correctly, I hope you do too. 
Screenshot 2024-08-06 at 19 04 34

**Review by latin-panda — APPROVED** [review]
Thanks Ben 
🤩 exciting to see this ready!! Nice job 🥳 🥳

**Comment by Benmuiruri** [comment]
Hi @tatilepizs 👋 . This is waiting your review when you have some time. Thanks.

**Review by tatilepizs — APPROVED** [review]
Looks really good @Benmuiruri, thank you! ⭐

**Pull request merged**